---
title: Classify T cells from multiple studies (human cancers)
author: "M. Andreatta <massimo.andreatta at unil.ch> and E. Prieto Fernandez <eprieto at cicbiogune.es>"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
  encoding=encoding,
  output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
#output: html_notebook
---

Libraries. Get all the necessary packages by 'restoring' the renv environment.
```{r warning=F, message=F, results=F}
library(renv)
renv::restore()

library(ggplot2)
library(Seurat)
library(UCell)
library(scGate)
library(Biostrings)
library(ProjecTILs)
library(SingleCellExperiment)
library(viridis)
library(RColorBrewer)
library(msigdbr)
library(presto)
library(tibble)
library(dplyr)
library(fgsea)
library(corrplot)
library(ggpubr)
library(xlsx)
library(pheatmap)
library(stringr)
library(ggrepel)
```

We will use the collection of single-cell data by [Zheng et al. (2021) Science](https://www.science.org/doi/10.1126/science.abe6474), which covers 21 cancer types and 300 patients.

The data can be downloaded from Zenodo: https://doi.org/10.5281/zenodo.5461803
Direct download link: https://zenodo.org/record/5461803/files/data.expression.tar.gz?download=1

```{r}
#After data download, set the path below to point to your data folder
datapath <- "~/Dropbox/CSI/Datasets/Zhang_TIL_Science2021/expression/CD8/byDataset"
#datapath <- "./Zhang_TIL_Science2021/expression/CD8/byDataset"


data.list <- list()
for (file in list.files(datapath)) {
  name <- gsub(pattern='.sce.rds', replacement = '', x=file)
  data.list[[name]] <- readRDS(sprintf("%s/%s",datapath,file))
}
names(data.list)
```

Convert to Seurat object
```{r warning=F}
seurat.list <- lapply(X = seq_along(data.list),  
                    FUN=function(i) {
                        mainExpName(data.list[[i]]) <- 'RNA'
                        study.name <- names(data.list)[i]
                        study.id <- sprintf("S%02d",i)
                        
                        a <- names(assays(data.list[[i]]))
                        if ("counts" %in% a) {
                           counts.slot = "counts"
                        } else if ("log2TPM" %in% a) {
                           counts.slot = "log2TPM"
                        } else if ("norm_exprs" %in% a) {
                           counts.slot = "norm_exprs" 
                        }  
                        
                        #Use gene names for all objects (stored in display.name)
                        rownames(data.list[[i]]) <- rowData(data.list[[i]])[["display.name"]]  

                        r <- as.Seurat(data.list[[i]], counts = counts.slot, data = "norm_exprs", project=study.name)
                        r <- RenameCells(r, add.cell.id = study.id)
                        
                        r
                    })
names(seurat.list) <- names(data.list)

rm(data.list)  #free up memory
gc()
```

Check distributions of normalized counts
```{r fig.height=5, fig.width=8}
gene <- "CD2"
pll <- lapply(names(seurat.list),
       function(x){
          df <- data.frame(counts = seurat.list[[x]]@assays$RNA@counts[gene,])
          ggplot(df, aes(x=counts)) + geom_histogram(bins=50) + ggtitle(x)
       })
wrap_plots(pll)

pll <- lapply(names(seurat.list),
       function(x){
          df <- data.frame(norm.data = seurat.list[[x]]@assays$RNA@data[gene,])
          ggplot(df, aes(x=norm.data)) + geom_histogram(bins=50) + ggtitle(x)
       })
wrap_plots(pll)
```

Data seems to be normalized in inconsistent way. Most datasets have expression of CD8A of ~3 on average while other subsets of ~9 on average. Try to correct by measuring 90% percentile of key genes.

Re-scale flagged datasets
```{r  fig.height=8, fig.width=12}
ref.factor <- 3
seurat.list <- lapply(seurat.list,
       function(x){
          df <- data.frame(x@assays$RNA@data[c("CD2","CD3E","CD3G","CD3D","CD8A","CD8B","PTPRC"),])
          q90 <- apply(df, 1, function(i){quantile(i, c(0.9))[1]})
          q90.mean <- mean(q90)
          print(sprintf("%s - %.3f", unique(x$dataset)[1], q90.mean))
          
          factor <- ref.factor / q90.mean
          #Rescale data with correction factor
          x@assays$RNA@data <- factor * x@assays$RNA@data
          x
       })
```

```{r fig.height=5, fig.width=8}
pll <- lapply(names(seurat.list),
       function(x){
          df <- data.frame(norm.data = seurat.list[[x]]@assays$RNA@data["CD8A",])
          ggplot(df, aes(x=norm.data)) + geom_histogram(bins=50) + ggtitle(x)
       })
wrap_plots(pll)
```

Save intermediate object
```{r}
dir.create('aux', showWarnings = FALSE)
dir.create('plots', showWarnings = FALSE)

saveRDS(seurat.list, "aux/Zheng.seurat.list.rds")
#seurat.list <- readRDS("aux/Zheng.seurat.list.rds")
```

Check some signatures for major cell types
```{r}
sigs <- list("Immune"=c("PTPRC"),
             "Myeloid"=c("SPI1"),
             "Tcell"=c("CD3D","CD3E","CD3G","CD2"),
             "Cd8T"=c("CD8A","CD8B","CD4-"),
             "Cd4T"=c("CD4","CD40LG"),
             "Treg"=c("CD4","FOXP3"),
             "Stromal"= c("MMP2","COL1A1","COL1A2","COL5A1","LUM","PDGFRA"),
             "Bcell" = c("CD79A","MS4A1","BANK1","PAX5","CD19"),
             "NK" = c("KLRD1","NKG7","NCR1","FCGR3A","CD3D-","CD3E-","CD3G-"),
             "MoMacDC" = c("LYZ","CSF1R","MSR1","MAFB","CD300E"),
             "Plasma_cell" = c("IGKC","IGHG3","IGHG1","IGHA1","CD19-"),
             "MAIT"=c("KLRB1", "SLC4A10", "NCR3"),
             "iNKT"=c("ZBTB16"),
             "KIR_NK"=c("TXK","KIR2DL3")
             )
cycling.g1 <- scGate::genes.blacklist.default$Hs$cellCycle.G1S
cycling.g2 <- scGate::genes.blacklist.default$Hs$cellCycle.G2M
cycling <- c(cycling.g1, cycling.g2)
sigs$cycling = cycling
seurat.list <- lapply(seurat.list,
       FUN = function(x) {
          AddModuleScore_UCell(x, features = sigs, name=NULL, ncores=4)
       })
```

```{r fig.height=6}
for (type in names(sigs)) {
  pll <- lapply(names(seurat.list),
                FUN = function(x) {
                  cols = c("lightgrey", "blue")
                  if (sum(seurat.list[[x]][[type]])==0) {
                     cols = c("lightgrey", "lightgrey")
                  }
                  
                  FeaturePlot(seurat.list[[x]], features=c(type), max.cutoff = 'q99', cols=cols) + 
                    ggtitle(sprintf("%s (%s)", x, type))
                })
  wrap_plots(pll)
  ggsave(sprintf("plots/Zheng.sets.UCell.scores.%s.png", type), height=18, width=25)
}
```

Filter CD8+ T cells
```{r warning=F, message=F}
models <- scGate::get_scGateDB()
cd8.model <- models$human$generic$CD8T

#Add MAITs to model
cd8.model <- scGate::gating_model(cd8.model, level=4, negative=T, name="MAIT", signature = c("KLRB1", "SLC4A10", "NCR3"))

gated.list <- lapply(names(seurat.list),
       FUN = function(x) {
          print(sprintf("#Running dataset %s", x))
          scGate(seurat.list[[x]], model=cd8.model, ncores=4, pca.dim = 30)
       })
names(gated.list) <- names(seurat.list)
rm(seurat.list)  #free up memory
gc()
```

Show results of gating per dataset
```{r fig.height=6}
cols = c("#10f87b","#fa765e")
names(cols) <- c("Pure","Impure")
pll <- lapply(names(gated.list),
       FUN = function(x) {
          DimPlot(gated.list[[x]], group.by="is.pure", cols = cols) + ggtitle(x)
       })
wrap_plots(pll)
ggsave("plots/Zheng.sets.gated.cd8t.png", height=18, width=25)
```

Show original annotations
```{r fig.height=6, fig.width=12}
cc <- lapply(gated.list, function(x) {
  unique(x$meta.cluster.coarse)
  })
clusters <- unique(unlist(cc))
palette <- brewer.pal(n = 12, name = 'Paired')
names(palette) <- clusters
pll <- lapply(names(gated.list),
       FUN = function(x) {
          DimPlot(gated.list[[x]], group.by="meta.cluster.coarse", cols = palette) + ggtitle(x)
       })
wrap_plots(pll)
ggsave("plots/Zheng.sets.by.coarse.cluster.png", height=18, width=35)
```

Plot by sample/patient
```{r fig.height=10, fig.width=20}
pll <- lapply(names(gated.list),
       FUN = function(x) {
          DimPlot(gated.list[[x]], group.by="patient") + ggtitle(x)
       })
wrap_plots(pll)
ggsave("plots/Zheng.sets.by.patient.png", height=18, width=35)
```

Save results of scGate
```{r}
saveRDS(gated.list, "aux/Zheng.seurat.gated.rds")
#gated.list <- readRDS("aux/Zheng.seurat.gated.rds")
```

Select pure cells and sufficiently large samples
```{r}
min.cells <- 200
large.samples <- lapply(names(gated.list),
       FUN = function(x) {
          tmp <- subset(gated.list[[x]], subset=is.pure=="Pure")
          tmp <- DietSeurat(tmp, dimreducs = NULL, assays = "RNA")
          
          ss <- SplitObject(tmp, split.by = 'patient')
          names(ss) <- paste0(x, "_", names(ss))
          
          ncells <- lapply(ss, ncol)
          ss <- ss[ncells>min.cells]
          
          ss
       })
large.samples <- unlist(large.samples, recursive = F)
names(large.samples)
```

Project individual samples in reference TIL atlas (ProjecTILs)
```{r}
ref <- load.reference.map()
#Focus on CD8+ T cells
is.cd8 <- colnames(ref)[ref$functional.cluster %in% c("CD8_Tex","CD8_Tpex","CD8_EffectorMemory","CD8_EarlyActiv","CD8_NaiveLike")]
ref.cd8 <- subset(ref, cells=is.cd8)
```

Make projection
```{r}
#ref.use = ref
ref.use = ref.cd8
projected <- make.projection(large.samples, ref=ref.use, filter.cells = FALSE, skip.normalize = T, ncores=8)
projected <- lapply(projected, function(x) {
                      cellstate.predict(ref=ref.use, query=x)
                    })
plot.projection(ref.cd8)
```

Save results
```{r}
saveRDS(projected, "aux/Zheng.seurat.projected.rds")
#projected <- readRDS("aux/Zheng.seurat.projected.rds")
```

Plot projections by study
```{r fig.height=30, fig.width=40}
pll <- lapply(names(projected),
       FUN = function(x) {
          plot.projection(projected[[x]], ref=ref.use) + NoLegend() + ggtitle(x)
       })
wrap_plots(pll)
ggsave("plots/Zheng.samples.projected.png", height=35, width=50, limitsize = FALSE)
```

```{r fig.height=4, fig.width=10}
genes <- c("Tcf7","Ccr7","Il7r","Cd4","Cd8a","Gzmb","Gzmk","Tox","Pdcd1","Lag3","Havcr2","Cx3cr1","Foxp3")
ProjecTILs::plot.states.radar(ref=ref.use, query=projected[1:4], genes4radar = genes, min.cells = 50)
```

Focus on 10x/in drop technologies to have more homogeneous and comparable gene expression across datasets.
Also exclude the Melanoma.MosheSade-Feldman2018 study because some samples come from LN but all are marked as tumor.

```{r}
studies <- names(gated.list)

LN.samples <- "MELA.MosheSade-Feldman2018"
TPM.tech <- c("CHOL.thisStudy", "CRC.LeiZhang2018", "HCC.ChunhongZheng2017", "HCC.QimingZhang2019_SS2", "MELA.LivnatJerby-Arnon2018", "LC.XinyiGuo2018")

studies <- setdiff(studies, LN.samples)
studies <- setdiff(studies, TPM.tech)
```

Merge results by study to visualize jointly 
```{r}
matches <- lapply(studies,
                  function(s) {
                    match <- grep(pattern=s, x=names(projected))
                    match
                    })
names(matches) <- studies
matches <- matches[lapply(matches,length)>0]
projected.by.study <- lapply(names(matches),
                             function(s) {
                               x <- matches[[s]]
                               print(sprintf("%s - merging %i sets",s, length(x)))
                               if (length(x) == 1) {
                                  projected[[x]]
                               } else {
                                  Reduce(f = ProjecTILs:::merge.Seurat.embeddings, x = projected[x])
                               }
                             })
names(projected.by.study) <- names(matches)
length(projected.by.study)                        
names(projected.by.study)
rm(projected)  #free up memory
gc()
```

Plot projections by study
```{r fig.height=15, fig.width=20}
pll <- lapply(names(projected.by.study),
       FUN = function(x) {
          plot.projection(projected.by.study[[x]], ref=ref.use, pointsize = 0.5, linesize = 0.5) + NoLegend() + ggtitle(x)
       })
wrap_plots(pll)
ggsave("plots/Zheng.samples.projected.by.study.pdf", height=22, width=25)
```

Agreement between studies (radar plots) and with reference expression profiles
```{r fig.height=4, fig.width=10}
nb.cols <- length(matches)
my.colors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)

marker.genes <- c("Tcf7","Ccr7","Il7r","Cd4","Cd8a","Gzmb","Gzmk","Tox","Pdcd1","Lag3","Havcr2","Cx3cr1","Cxcl13")

proteasome.genes <- c("Psma1","Psma2","Psma3","Psma4","Psma5","Psma6","Psma7","Psmb1","Psmb2","Psmb3","Psmb4","Psmb5","Psmb6","Psmb7","Psmc1","Psmc2","Psmc3","Psmc4","Psmc5","Psmc6","Psmd1","Psmd2","Psmd3","Psmd4","Psmd6","Psmd7","Psmd8","Psmd11","Psmd12","Psmd13","Psmd14")
chaperones <- c("Psmg2", "Pomp", "Psmd9")

a <- ProjecTILs::plot.states.radar(ref=ref.use, query=projected.by.study[1:length(matches)], genes4radar = marker.genes, min.cells = 30, return = T)
ggsave("plots/marker_genes.pdf", plot = a, height=12, width=25)

b <- ProjecTILs::plot.states.radar(ref=ref.use, query=projected.by.study[1:length(matches)], genes4radar = proteasome.genes, min.cells = 30, return = T)
ggsave("plots/proteasome_genes.pdf", plot = a, height=12, width=25)

c <- ProjecTILs::plot.states.radar(ref=ref.use, query=projected.by.study[1:length(matches)], genes4radar = chaperones, min.cells = 30, return = T)
ggsave("plots/chaperones.pdf", plot = a, height=12, width=25)
```

Merge all and downsample to max 1000 cells per dataset
```{r}
ds <- 1000
projected.sub <- lapply(projected.by.study, function(x) {
  Idents(x) <- "dataset"
  subset(x, downsample = ds)
})
projected.merged.all <- Reduce(f = ProjecTILs:::merge.Seurat.embeddings, x = projected.sub)
```

Save results
```{r}
saveRDS(projected.merged.all, "aux/Zheng.seurat.projected.ds.rds")
#projected.merged.all <- readRDS("aux/Zheng.seurat.projected.ds.rds")
```

Focus on exhausted T cells (Tex) and effector-memory T cells (Tem)
```{r}
projected.merged.Temex <- subset(projected.merged.all,
                                 subset=functional.cluster %in% c("CD8_EffectorMemory","CD8_Tex"))
```

Save results
```{r}
saveRDS(projected.merged.Temex, "aux/Zheng.seurat.projected.TexTem.ds.rds")
#projected.merged.Temex <- readRDS("aux/Zheng.seurat.projected.TexTem.ds.rds")
```

View global embeddings
```{r}
palette <- c("#edbe2a", "#A58AFF", "#53B400", "#F8766D", "#00B6EB", "#d1cfcc", "#FF0000", "#87f6a5", "#e812dd")
names(palette) <- levels(factor(ref.use$functional.cluster))
DimPlot(projected.merged.all, reduction = "umap", cols=palette, group.by = "functional.cluster")
DefaultAssay(projected.merged.all) <- "RNA"
FeaturePlot(projected.merged.all, features=c("Zbtb16","Blk","Gzmk","Sell"), max.cutoff = 'q98')
```

Recalculate embeddings
```{r}
DefaultAssay(projected.merged.all) <- "integrated"
ndim <- dim(ref.use@misc$umap_object$data)[2]
projected.merged.all <- RunUMAP(projected.merged.all, dims = 1:ndim)
my.colors <- c("#F781BF","#5E3C99","gray70","#FDB863","lightseagreen")
DimPlot(projected.merged.all, reduction = "umap", pt.size = 0.75, cols=my.colors, group.by = "functional.cluster") + NoLegend()
ggsave("plots/FigS1b_legend.pdf", width = 7, height = 7)

DimPlot(projected.merged.all, reduction = "umap", group.by = "dataset", split.by="dataset", ncol = 6) + NoLegend()
```

Color by study
```{r}
nb.cols <- length(matches)
my.colors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)
DimPlot(projected.merged.all, reduction = "umap", pt.size = 0.75, cols=my.colors, group.by = "dataset") + NoLegend()
ggsave("plots/FigS1a.pdf", width = 7, height = 7)
```

What are the outlier clusters?
```{r}
npc <- 15
resol <- 0.3
DefaultAssay(projected.merged.all) <- "integrated"
projected.merged.all <- FindNeighbors(projected.merged.all, reduction = "pca", dims = 1:npc)
projected.merged.all <- FindClusters(projected.merged.all, resolution = resol)
projected.merged.all$cluster <- projected.merged.all@active.ident
DimPlot(projected.merged.all, group.by = "cluster", label = T) + NoLegend()
```

Cluster 5 (Zbtb16) may be iNKT cells
```{r}
DefaultAssay(projected.merged.all) <- 'RNA'
cluster.markers <- FindAllMarkers(projected.merged.all, only.pos = T, min.pct = 0.2, min.diff.pct=0.05, 
                                  logfc.threshold = 0.1, max.cells.per.ident = 1000, test.use="wilcox", base=exp(1))

all <- cluster.markers %>% group_by(cluster) %>% top_n(n = 50, wt = abs(avg_logFC))

for (i in levels(projected.merged.all@active.ident)) {
    print(subset(all, cluster==i))
}
```

Internal controls
Map functional cluster predictions back to original objects (work with human genes)

```{r}
studies.included <- names(projected.by.study)
gated.ann <- lapply(studies.included, function(n) {
    x <- gated.list[[n]]
    x$functional.cluster <- NA
    
    ann <- projected.by.study[[n]]$functional.cluster
    x$functional.cluster[names(ann)] <- ann
    
#    print(table(x$functional.cluster, useNA = "ifany"))
    x
})
names(gated.ann) <- names(projected.by.study)
```

Subset on Tex and T_EM only
```{r}
gated.sub <- lapply(gated.ann, function(x) {
    subset(x, subset=functional.cluster %in% c("CD8_EffectorMemory","CD8_Tex"))
})
saveRDS(gated.sub, "aux/Zheng.seurat.gated.annotated.Tex_Tem.rds")
```

Show ProjecTILs annotations
```{r fig.height=6, fig.width=12}
palette <- c("#edbe2a", "#A58AFF", "#53B400", "#F8766D", "#00B6EB")
names(palette) <- c("CD8_Tex","CD8_Tpex","CD8_EffectorMemory","CD8_EarlyActiv","CD8_NaiveLike")
pll <- lapply(names(gated.ann),
       FUN = function(x) {
          DimPlot(gated.ann[[x]], group.by="functional.cluster", cols = palette) + ggtitle(x)
       })

wrap_plots(pll)
ggsave("plots/Zheng.sets.by.functional.cluster.pdf", height=18, width=35)
```

```{r fig.height=5}
pll <- lapply(names(gated.ann),
       FUN = function(x) {
        tt <- t(table(zheng=gated.ann[[x]]$meta.cluster,projectils=gated.ann[[x]]$functional.cluster))
        ggplot(data = melt(tt), aes(x=zheng, y=projectils, fill=value)) + geom_tile()  + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
         
       })
wrap_plots(pll)
ggsave("plots/Zheng.vs.projectils.heatmaps.pdf", height=18, width=35)
```

Subset on Tex and Tem only
```{r}
gated.sub <- lapply(gated.ann, function(x) {
    subset(x, subset=functional.cluster %in% c("CD8_EffectorMemory","CD8_Tex"))
})
```

Save results
```{r}
saveRDS(gated.sub, "aux/Zheng.seurat.gated.annotated.Tex_Tem.rds")
#gated.sub <- readRDS("aux/Zheng.seurat.gated.annotated.Tex_Tem.rds")
```

Composition barplot
```{r fig.height=5}
pll <- lapply(names(gated.sub),
       FUN = function(x) {
        tt <- (table(zheng=gated.sub[[x]]$meta.cluster,projectils=gated.sub[[x]]$functional.cluster))
        tt <- scale(tt,center = F, scale = colSums(tt))
        tt <- melt(tt)
        p <- ggplot(tt, aes(x = projectils, y = value, fill = zheng)) + 
            geom_bar(stat = "identity") + theme_bw()  + 
            theme(legend.position = "left") + ggtitle(x)
       
       })
wrap_plots(pll)
ggsave("plots/Zheng.vs.projectils.BARPLOT.Tex_T_EM.pdf", height=25, width=35)
```

Additional internal controls
```{r fig.height=4}
typical.Texem <- c("Tox","Entpd1","Cxcl13","Ctla4","Havcr2","Lag3","Pdcd1","Il7r", "Tcf7", "Gzmk","Gzmm", "Anxa1", "Ccl5")

DefaultAssay(projected.merged.Temex) <- "RNA"
pll <- FeaturePlot(projected.merged.Temex, combine = F, max.cutoff = 'q98',
            features=typical.Texem) 
pll <- lapply(pll, function(x) {
   x + scale_color_viridis(option="D", direction=1)
})
wrap_plots(pll)

pll <- FeaturePlot(projected.merged.Temex, combine = F, max.cutoff = 'q98',
            features=typical.Texem) 

pll <- lapply(pll, function(x) {
   x + scale_color_viridis(option="D", direction=1, begin=0.2) + 
      theme(aspect.ratio = 1, axis.title=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank())
})
wrap_plots(pll)

ggsave("plots/markers.integrated.UMAP.pdf", height=20, width=20)
```

Piechart by study
```{r}
table <- table(projected.merged.all$dataset, projected.merged.all$functional.cluster)
labels <- rownames(table)
number <- as.vector(table[,1]) + as.vector(table[,2]) + as.vector(table[,3]) + as.vector(table[,4]) + as.vector(table[,5])

pdf("plots/FigS1a_pie_study.pdf")
pie(x = number, labels = labels, main="Contribution by study", col = my.colors, clockwise = T)
dev.off()
```

Piechart by cancer type
```{r}
table <- table(projected.merged.all$cancerType, projected.merged.all$functional.cluster)
labels <- rownames(table)
number <- as.vector(table[,1]) + as.vector(table[,2]) + as.vector(table[,3]) + as.vector(table[,4]) + as.vector(table[,5])
my.colors <- colorRampPalette(brewer.pal(8, "Paired"))(20)

pdf("plots/FigS1a_pie_cancerType.pdf")
pie(x = number, labels = labels, main="Contibution by cancerType", col = my.colors, clockwise = T)
dev.off()
```

Check phenotype markers
```{r fig.height=4}
DefaultAssay(projected.merged.all) <- "RNA"
phenotype.markers <- c("Cd8a","Cd4","Tcf7","Ccr7","Cd69","Gzmk","Gzmb","Havcr2","Lag3","Entpd1","Cxcr6","Mki67","Top2a","Anxa1","Gzmm","Zbtb16","Hivep3","Ifit3","Hsp90aa1","Sell")
pll <- FeaturePlot(projected.merged.all, combine = F, max.cutoff = 'q98',
            features=phenotype.markers) 

wrap_plots(pll)
ggsave("plots/FigS1a_phenotype markers.pdf", width = 20, height = 15)
```

View global embeddings of Tex and Tem cells
```{r}
palette <- c("#5E3C99","#FDB863")
DefaultAssay(projected.merged.Temex) <- "integrated"
ndim <- dim(ref.use@misc$umap_object$data)[2]
projected.merged.Temex <- RunUMAP(projected.merged.Temex, dims = 1:ndim)
DimPlot(projected.merged.Temex, reduction = "umap", cols=palette, pt.size = 0.75, group.by = "functional.cluster")  +
     ggtitle("CD8+ TILs") + theme(aspect.ratio = 1) + NoLegend()
DimPlot(projected.merged.Temex, reduction = "umap", group.by = "dataset", split.by="dataset", ncol = 6) + NoLegend()
ggsave("plots/Fig1a.pdf", width = 7, height = 7)
```
Save results
```{r}
saveRDS(projected.merged.Temex, "aux/Zheng.seurat.projected.TexTmex.ds.rds")
#projected.merged.Temex <- readRDS("aux/Zheng.seurat.projected.TexTmex.ds.rds")
```

Color by study TexTem
```{r}
nb.cols <- length(matches)
my.colors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)
DimPlot(projected.merged.Temex, reduction = "umap", pt.size = 0.75, cols=my.colors, group.by = "dataset") + NoLegend()
ggsave("plots/FigS1c.pdf", width = 7, height = 7)
```
Pie chart Tex by cancer type
```{r}
table <- table(projected.merged.Temex$cancerType, projected.merged.Temex$functional.cluster)
labels <- rownames(table)
number <- as.vector(table[,2])
my.colors <- colorRampPalette(brewer.pal(8, "Paired"))(20)

pdf("plots/FigS1c_cancerType_Tex.pdf")
pie(x = number, labels = labels, main="Tex Contribution by cancerType", col = my.colors, clockwise = T)
dev.off()
```

Pie chart Tem by cancer type
```{r}
table <- table(projected.merged.Temex$cancerType, projected.merged.Temex$functional.cluster)
labels <- rownames(table)
number <- as.vector(table[,1])
my.colors <- colorRampPalette(brewer.pal(8, "Paired"))(20)

pdf("plots/FigS1c_cancerType_Tex.pdf")
pie(x = number, labels = labels, main="Tex Contribution by cancerType", col = my.colors, clockwise = T)
dev.off()
```

Expression of typical Tex/Tem markers
```{r}
Idents(projected.merged.Temex) <- "functional.cluster"
VlnPlot(projected.merged.Temex, features= typical.Texem, pt.size = 0, stack = T, flip = F, cols=palette, fill.by = "ident")
ggsave("plots/FigS2a.pdf", height=3, width=12)
```

# Make heatmap function
```{r}
nb.cols <- length(matches)
my.colors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)

make.heatmap <- function(data, assay="RNA", genes, method=c("ward.D", "average"), brewer.palette="PRGn") {
  
  library(pheatmap)
  set.seed(123)
  #Calculate mean expression by subset
  m <- c()
  for( g in unique(genes)){
    m[[g]] <- tapply(data@assays[[assay]][g,],data$subset_dataset_patient, mean)
  }
  m <- as.data.frame(m)
  m.study <- factor(unlist(lapply(strsplit(rownames(m),"@",perl = T),function(x) x[[2]])))
  m.subset <- factor(unlist(lapply(strsplit(rownames(m),"@",perl = T),function(x) x[[1]])))
  breaksList = seq(-2, 2, by = 0.1)
  color = colorRampPalette(rev(brewer.pal(n = 7, name = brewer.palette)))(length(breaksList))
  gaps_col <- (which(!duplicated(m.subset))-1)[-1]
  annotation_col = data.frame(
    Study = m.study, 
    TIL_Subtype = m.subset
  )
  rownames(annotation_col) = rownames(m)
  palette <- c("#5E3C99","#FDB863")
  palette2 <- my.colors
  names(palette) <- levels(m.subset)
  names(palette2) <- levels(m.study)
  h <- pheatmap::pheatmap(t(m),cluster_rows = F,
                          cluster_cols = F,scale = "row",
                          breaks = breaksList, color=color, 
                          annotation_col = annotation_col, 
                          gaps_col=gaps_col,show_colnames = F,
                          annotation_colors = list(TIL_Subtype=palette, Study=palette2), 
                          fontsize_row=6,fontsize = 7, 
                          clustering_method=method)
  
  return(h)
}
```

Heatmap Texem
```{r}
data.heat <- projected.merged.Temex
DefaultAssay(data.heat) <- "RNA"
data.heat$dataset_patient <- paste0(data.heat$dataset, "@", data.heat$patient)
data.heat$subset_dataset_patient <- paste0(data.heat$functional.cluster, "@", data.heat$dataset_patient)
h <- make.heatmap(data.heat, genes=typical.Texem, brewer.palette = "PiYG")
ggsave("plots/Fig1b.pdf", plot=h, height=3, width=10)
```

Expression of proteasomal genes
```{r}

Idents(projected.merged.Temex) <- "functional.cluster"
VlnPlot(projected.merged.Temex, features=proteasome.genes, pt.size = 0, stack = T, flip = F, cols=palette, fill.by = "ident")
ggsave("plots/FigS2c.pdf", height=3, width=18)

```

Expression of chaperones
```{r}

Idents(projected.merged.Temex) <- "functional.cluster"
VlnPlot(projected.merged.Temex, features=chaperones, pt.size = 0, stack = T, flip = F, cols=palette, fill.by = "ident")
ggsave("plots/FigS2d.pdf", height=3, width=6)

```

Heatmap proteasome genes
```{r}
data.heat <- projected.merged.Temex
DefaultAssay(data.heat) <- "RNA"
data.heat$dataset_patient <- paste0(data.heat$dataset, "@", data.heat$patient)
data.heat$subset_dataset_patient <- paste0(data.heat$functional.cluster, "@", data.heat$dataset_patient)
h <- make.heatmap(data.heat, genes=proteasome.genes, brewer.palette = "PiYG")
ggsave("plots/Fig1e.pdf", plot=h, height=4, width=10)
```

# GSEA analysis
```{r}
markers.genes <- wilcoxauc(X = projected.merged.Temex, 'functional.cluster')
markers.genes$log10p <- -log10(markers.genes$padj)
dplyr::count(markers.genes, group)
markers.genes <- markers.genes %>%
  dplyr::filter(padj <= 0.05) %>%
  dplyr::filter(group == "CD8_Tex") %>%
  arrange(desc(logFC)) %>%
  dplyr::select(feature, logFC,log10p)
write.xlsx(markers.genes, "DE genes.xlsx", sheetName = "padj", append = TRUE)
ranks<- deframe(markers.genes)
head(ranks)
barplot(sort(ranks, decreasing = T))
```

Volcano plot
```{r}
markers.genes <- wilcoxauc(X = projected.merged.Temex, 'functional.cluster')
markers.genes$log10p <- -log10(markers.genes$padj)
dplyr::count(markers.genes, group)
markers.genes <- markers.genes %>%
  #dplyr::filter(padj <= 0.05) %>%
  dplyr::filter(group == "CD8_Tex") %>%
  arrange(desc(logFC)) %>%
  dplyr::select(feature, logFC,log10p)

'%!in%' <- function(x,y)!('%in%'(x,y))

proteasome.signature <- c("Psmb3", "Psma2", "Psmb2", "Psma4", "Psmd8", "Psmc3", "Psmb9", "Pomp", "Psmb1", "Psma1", "Psma7")
exhaustion = c("Tox", "Entpd1", "Cxcl13", "Ctla4","Havcr2", "Lag3", "Pdcd1")
effector.memory = c("Il7r","Tcf7","Gzmk","Gzmm","Anxa1","Ccl5")

volcano <- markers.genes %>% 
  mutate(
    cond = case_when( 
      logFC >= 0.2 & log10p >= 1.30103 & feature %in% proteasome.signature ~ "blue",
      logFC >= 0.2 & log10p >= 1.30103 & feature %!in% proteasome.signature ~ "#FDB863",
      logFC < -0.2 & log10p >= 1.30103 ~ "#5E3C99",
      TRUE ~ "gray70"),
    order = case_when(
      feature %in% proteasome.signature ~ "2",
      TRUE ~ "1"
      )
  )
volcano <- arrange(volcano, order)
labels <- c(effector.memory, exhaustion, proteasome.signature)
pll <- ggplot(volcano, aes(x= logFC, y= log10p)) + 
  geom_point(color= volcano$cond) +
  xlab(bquote(~log[2]~'FC')) + 
  xlim(c(-1.15,1.15)) +
  ylim(c(0,300)) +
  ylab(bquote(~-log[10]~'padj')) +
  geom_text_repel(aes(label=ifelse(feature %in% labels, as.character(feature), '')),
                  hjust=1.25,
                  fontface= 3,
                  max.overlaps = 8000) +
  geom_hline(yintercept=-log10(0.05),linetype="dashed",col="black") +
  geom_vline(xintercept=log2(1.148698),linetype="dashed",col="black") +
  geom_vline(xintercept=log2(1/1.148698),linetype="dashed",col="black") +
  theme(axis.line = element_line(color='black', size=0.5), panel.border = element_rect(color='black', size=0.5, fill=NA),panel.background = NULL)

ggsave("plots/volcano.pdf", plot=pll)
```

# Gene set enrichment analysis (GSEA)

In order to perform the GSEA the annotated gene set is needed. One popular source is the MsigDB from Broad Institute.
https://www.gsea-msigdb.org/gsea/msigdb/index.jsp H/C2/C5/C7
```{r}
m_df<- msigdbr(species = "Homo sapiens", category = "C2")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
names(ranks) <- toupper(names(ranks))
fgseaRes<- fgsea(fgsea_sets, ranks, minSize=15, maxSize = 500, nperm= 1000)
head(fgseaRes[order(padj, -abs(NES)), ], n=10)
write.xlsx(fgseaRes[,c(1:7)], "GSEA_pathways.xlsx", sheetName = "C2", append = TRUE)

m_df<- msigdbr(species = "Homo sapiens", category = "C5")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
names(ranks) <- toupper(names(ranks))
fgseaRes<- fgsea(fgsea_sets, ranks, minSize=15, maxSize = 500, nperm= 1000)
head(fgseaRes[order(padj, -abs(NES)), ], n=10)
write.xlsx(fgseaRes[,c(1:7)], "GSEA_pathways.xlsx", sheetName = "C5", append = TRUE)

m_df<- msigdbr(species = "Homo sapiens", category = "C7")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
names(ranks) <- toupper(names(ranks))
fgseaRes<- fgsea(fgsea_sets, ranks, minSize=15, maxSize = 500, nperm= 1000)
head(fgseaRes[order(padj, -abs(NES)), ], n=10)
write.xlsx(fgseaRes[,c(1:7)], "GSEA_pathways.xlsx", sheetName = "C7", append = TRUE)

m_df<- msigdbr(species = "Homo sapiens", category = "H")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
names(ranks) <- toupper(names(ranks))
fgseaRes<- fgsea(fgsea_sets, ranks, minSize=15, maxSize = 500, nperm= 1000)
head(fgseaRes[order(padj, -abs(NES)), ], n=10)
write.xlsx(fgseaRes[,c(1:7)], "GSEA_pathways.xlsx", sheetName = "H", append = TRUE)
```

Gene signatures from GSEA pathways
```{r}
m_df<- msigdbr(species = "Homo sapiens", category = "C2")
fgsea_sets.2 <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
m_df<- msigdbr(species = "Homo sapiens", category = "C5")
fgsea_sets.5 <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
m_df<- msigdbr(species = "Homo sapiens", category = "C7")
fgsea_sets.7 <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
```

Calculation of UCell signature scores
```{r}
signatures <- list(PROTEASOME_DEBLAS = proteasome.signature,
                   PROTEASOME = proteasome.genes,
                   T_EX = exhaustion,
                   T_EM = effector.memory,
                   BIOCARTA_PROTEASOME_PATHWAY = str_to_title(fgsea_sets.2$BIOCARTA_PROTEASOME_PATHWAY),
                   KEGG_PROTEASOME = str_to_title(fgsea_sets.2$KEGG_PROTEASOME),
                   REACTOME_CELLULAR_RESPONSE_TO_HYPOXIA = str_to_title(fgsea_sets.2$REACTOME_CELLULAR_RESPONSE_TO_HYPOXIA),
                   REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES = str_to_title(fgsea_sets.2$REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES),
                   GOCC_PROTEASOME_COMPLEX = str_to_title(fgsea_sets.5$GOCC_PROTEASOME_COMPLEX),
                   GOBP_CELLULAR_OXIDANT_DETOXIFICATION = str_to_title(fgsea_sets.5$GOBP_CELLULAR_OXIDANT_DETOXIFICATION),
                   GOBP_REGULATION_OF_PROTEASOMAL_PROTEIN_CATABOLIC_PROCESS = str_to_title(fgsea_sets.5$GOBP_REGULATION_OF_PROTEASOMAL_PROTEIN_CATABOLIC_PROCESS),
                   GOBP_REGULATION_OF_RESPONSE_TO_OXIDATIVE_STRESS = str_to_title(fgsea_sets.5$GOBP_REGULATION_OF_RESPONSE_TO_OXIDATIVE_STRESS),
                  
                   GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_DN = str_to_title(fgsea_sets.7$GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_DN),
                   GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_UP = str_to_title(fgsea_sets.7$GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_UP))

DefaultAssay(projected.merged.Temex) <- "RNA"
data_seurat <- AddModuleScore_UCell(projected.merged.Temex, features = signatures)
signature.names <- paste0(names(signatures), "_UCell")
pll <- VlnPlot(data_seurat, features = signature.names, group.by = "functional.cluster")
```

AUC calculation
```{r}
myFeaturesUcell <- paste0(names(signatures),"_UCell")
selFeatures <- myFeaturesUcell
data_seurat$T_EX_sub_T_EM_UCell <- data_seurat$T_EX_UCell - data_seurat$T_EM_UCell
selFeatures <- c(myFeaturesUcell,"T_EX_sub_T_EM_UCell")
w <- wilcoxauc(t(data_seurat@meta.data[,selFeatures]),y=data_seurat$functional.cluster)
ww <- w %>% filter(group == "CD8_Tex") %>% arrange(desc(auc)) %>% select(feature, logFC, auc, padj) 
ww
write.xlsx2(ww,"AUC.xlsx", sheetName = "CD8_Tex", append = TRUE)

pll <- ggplot(ww[c(2,4, 6, 18),], aes(x=reorder(feature,-auc), y=auc)) + geom_bar(stat="identity") + theme_bw() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1), 
        axis.title.x=element_blank()) +
  coord_cartesian(ylim=c(0.5, 0.8)) +
  ggtitle("Prediction of T cell subtypes")
pll
ggsave("plots/Fig1f.pdf", plot=pll, height=7, width=7)
```

Correlation plots
```{r}

cor.plot <- as.data.frame(data_seurat$PROTEASOME_DEBLAS_UCell)
cor.plot$T_EX_sub_T_EM_UCell <- data_seurat$T_EX_sub_T_EM_UCell
cor.plot$T_EX_UCell <- data_seurat$T_EX_UCell
cor.plot$REACTOME_CELLULAR_RESPONSE_TO_HYPOXIA_UCell <- data_seurat$REACTOME_CELLULAR_RESPONSE_TO_HYPOXIA_UCell
cor.plot$GOBP_CELLULAR_OXIDANT_DETOXIFICATION_UCell <- data_seurat$GOBP_CELLULAR_OXIDANT_DETOXIFICATION_UCell
cor.plot$REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES_UCell <- data_seurat$REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES_UCell
cor.plot$GOBP_REGULATION_OF_RESPONSE_TO_OXIDATIVE_STRESS_UCell <- data_seurat$GOBP_REGULATION_OF_RESPONSE_TO_OXIDATIVE_STRESS_UCell
cor.plot$KEGG_PROTEASOME_UCell <- data_seurat$KEGG_PROTEASOME_UCell
cor.plot$GOCC_PROTEASOME_COMPLEX_UCell <- data_seurat$GOCC_PROTEASOME_COMPLEX_UCell
cor.plot$PROTEASOME_UCell <- data_seurat$PROTEASOME_UCell
cor.plot$functional.cluster <- data_seurat$functional.cluster
colnames(cor.plot)

# Proteasome vs Detoxification of ROS
pll <- ggscatter(cor.plot, x= "GOCC_PROTEASOME_COMPLEX_UCell", y = "REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES_UCell",
          add = "reg.line", conf.int = TRUE, combine= TRUE,
          #facet.by = "functional.cluster",
          cor.coef = FALSE, cor.method = "pearson",
          add.params = list(color = "black", fill = "functional.cluster"),
          xlab = "GOCC_PROTEASOME_COMPLEX_UCell", ylab = "REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES_UCell",
          color = "functional.cluster",
          palette = c(CD8_EffectorMemory = "#5E3C99", CD8_Tex= "#FDB863"),
          size = 2,
          rug= TRUE,
          fullrange = TRUE) + stat_cor(aes(color = functional.cluster))
pll
ggsave("plots/Proteasome_vs_DetoxificationROS.pdf", plot=pll, height=7, width=7)

# Proteasome_deBlas vs Detoxification of ROS
pll <- ggscatter(cor.plot, x= "data_seurat$PROTEASOME_DEBLAS_UCell", y = "REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES_UCell",
          add = "reg.line", conf.int = TRUE, combine= TRUE,
          #facet.by = "functional.cluster",
          cor.coef = FALSE, cor.method = "pearson",
          add.params = list(color = "black", fill = "functional.cluster"),
          xlab = "data_seurat$PROTEASOME_DEBLAS_UCell", ylab = "REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES_UCell",
          color = "functional.cluster",
          palette = c(CD8_EffectorMemory = "#5E3C99", CD8_Tex= "#FDB863"),
          size = 2,
          rug= TRUE,
          fullrange = TRUE) + stat_cor(aes(color = functional.cluster))
pll
ggsave("plots/Proteasome_dB_vs_DetoxificationROS.pdf", plot=pll, height=7, width=7)

# Proteasome vs Celullar ROS
pll <- ggscatter(cor.plot, x= "GOCC_PROTEASOME_COMPLEX_UCell", y = "GOBP_CELLULAR_OXIDANT_DETOXIFICATION_UCell",
          add = "reg.line", conf.int = TRUE, combine= TRUE,
          #facet.by = "functional.cluster",
          cor.coef = FALSE, cor.method = "pearson",
          add.params = list(color = "black", fill = "functional.cluster"),
          xlab = "GOCC_PROTEASOME_COMPLEX_UCell", ylab = "GOBP_CELLULAR_OXIDANT_DETOXIFICATION_UCell",
          color = "functional.cluster",
          palette = c(CD8_EffectorMemory = "#5E3C99", CD8_Tex= "#FDB863"),
          size = 1.5,
          shape= 3,
          rug= TRUE,
          fullrange = TRUE) + stat_cor(aes(color = functional.cluster))
pll
ggsave("plots/Proteasome_vs_Celullar_ROS.pdf", plot=pll, height=7, width=7)
```

GSEA plots
```{r}
m_df<- msigdbr(species = "Homo sapiens", category = "C2")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
pathway <- "REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES"

plotEnrichment(toupper(fgsea_sets[[pathway]]), ranks, ticksSize = 0.5, stats = T) +
  labs(title=pathway) +
  plotGseaTable(fgsea_sets[[pathway]], ranks, fgseaRes, gseaParam = 1) + 
  theme_bw()
ggsave("plots/REACTOME_DETOXIFICATION_OF_REACTIVE_OXYGEN_SPECIES.pdf", height=2.5, width=6)

m_df<- msigdbr(species = "Homo sapiens", category = "C2")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
pathway <- "REACTOME_CELLULAR_RESPONSE_TO_HYPOXIA"
plotEnrichment(toupper(fgsea_sets[[pathway]]), ranks, ticksSize = 0.5, stats = T) +
  labs(title=pathway) +
  plotGseaTable(fgsea_sets[[pathway]], ranks, fgseaRes, gseaParam = 1) + 
  theme_bw()
ggsave("plots/REACTOME_CELLULAR_RESPONSE_TO_HYPOXIA.pdf", height=2.5, width=6)

m_df<- msigdbr(species = "Homo sapiens", category = "C5")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
pathway <- "GOCC_PROTEASOME_COMPLEX"
plotEnrichment(toupper(fgsea_sets[[pathway]]), ranks, ticksSize = 0.5, stats = T) +
  labs(title=pathway) +
  plotGseaTable(fgsea_sets[[pathway]], ranks, fgseaRes, gseaParam = 1) + 
  theme_bw()
ggsave("plots/GOCC_PROTEASOME_COMPLEX.pdf", height=2.5, width=6)

m_df<- msigdbr(species = "Homo sapiens", category = "C5")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
pathway <- "GOBP_CELLULAR_OXIDANT_DETOXIFICATION"
plotEnrichment(toupper(fgsea_sets[[pathway]]), ranks, ticksSize = 0.5, stats = T) +
  labs(title=pathway) +
  plotGseaTable(fgsea_sets[[pathway]], ranks, fgseaRes, gseaParam = 1) + 
  theme_bw()
ggsave("plots/GOBP_CELLULAR_OXIDANT_DETOXIFICATION.pdf", height=2.5, width=6)

m_df<- msigdbr(species = "Homo sapiens", category = "C7")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
pathway <- "GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_UP"
plotEnrichment(toupper(fgsea_sets[[pathway]]), ranks, ticksSize = 0.5, stats = T) +
  labs(title=pathway) +
  plotGseaTable(fgsea_sets[[pathway]], ranks, fgseaRes, gseaParam = 1) + 
  theme_bw()
ggsave("plots/GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_UP.pdf", height=2.5, width=6)

m_df<- msigdbr(species = "Homo sapiens", category = "C7")
head(m_df)
fgsea_sets <- m_df %>% split(x = .$gene_symbol, f = .$gs_name)
pathway <- "GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_DN"
plotEnrichment(toupper(fgsea_sets[[pathway]]), ranks, ticksSize = 0.5, stats = T) +
  labs(title=pathway) +
  plotGseaTable(fgsea_sets[[pathway]], ranks, fgseaRes, gseaParam = 1) + 
  theme_bw()
ggsave("plots/GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_DN.pdf", height=2.5, width=6)
```


