---
title: Analyse proteasomal gene expression in Tex and T_EM
author: "M. Andreatta <massimo.andreatta at unil.ch> and S. Carmona <santiago.carmona at unil.ch>"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
#output: html_notebook
---

```{r}
library(ggplot2)
library(Seurat)
library(UCell)
library(scGate)
library(EnhancedVolcano)
library(Scillus)
library(msigdbr)
library(presto)
library(tibble)
library(fgsea)
```

El punto de partida es Tex_EffectorMemory_CD8.rds, que contiene las T cells previamente clasificadas como Tex o EffectorMemory por ProjecTILs (merged Seurat objects).

#NOTE: Can you share the scripts used to classify T cells from original single-cell data?
Starting point: several .rds files from which I extracted only the CD8_Tex and CD8_EffectorMemory subpopulations (ovarian, liver, breast, lung_1 & lung_2)

```{r}
#merged <- merge(x=lung_2, y= c(lung_1, breast, liver, ovarian))
#DefaultAssay(merged) <- "RNA"

#merged <- FindVariableFeatures(merged, selection.method = "vst", nfeatures = 3000, verbose = T)
#merged <- ScaleData(merged, verbose = FALSE)
#merged <- RunPCA(merged, npcs = 20, verbose = FALSE)
#merged <- RunUMAP(merged, reduction = "pca", dims = 1:20)
#merged <- FindNeighbors(merged, reduction = "pca", dims = 1:20)
#merged <- FindClusters(merged, resolution = 2.0)

#saveRDS(merged, "Pan_Tex_EffectorMemory_CD8.rds", refhook = NULL)
```

###############################################################################################
######### Load the Seurat object and remove lung_1 (excluded  after heatmap analysis) #########
###############################################################################################

```{r}
ddir <- "./data"
merged <- readRDS(sprintf("%s/Tex_EffectorMemory_CD8.rds", ddir))
merged
```

Have a look of the data
```{r}
DimPlot(merged)
DimPlot(merged, group.by = "cancer.type")
DimPlot(merged, group.by = "functional.cluster")

FeaturePlot(merged, features=c("Cd2","Cd8a","Cd8b1","Cd4"))
FeaturePlot(merged, features=c("Lck","Spi1","Ms4a1","Igha"))
```

Check some signatures for major cell types
```{r}
sigs <- list("Immune"=c("Ptprc"),
             "Myeloid"=c("Spi1"),
             "Tcell"=c("Cd3d","Cd3e","Cd3g","Cd2"),
             "Cd8T"=c("Cd3d","Cd3e","Cd3g","Cd2","Cd8a","Cd8b1","Cd4-"),
             "Stromal"= c("Mmp2","Col1a1","Col1a2","Col5a1","Lum","Pdgfra"),
             "Bcell" = c("Ms4a1","Bank1","Pax5","Cd19"),
             "NK" = c("Klrd1","Nkg7","Ncr1","Cd3d-","Cd3e-"),
             "MoMacDC" = c("Lyz1","Lyz2","Csf1r","Msr1","Mafb","Cd300e")
             )

cycling.g1 <- scGate::genes.blacklist.default$Mm$cellCycle.G1S
cycling.g2 <- scGate::genes.blacklist.default$Mm$cellCycle.G2M
cycling <- c(cycling.g1, cycling.g2)

sigs$cycling = cycling

merged <- AddModuleScore_UCell(merged, features = sigs, name=NULL, ncores=4)

```

```{r}
FeaturePlot(merged, features=c("Immune","Myeloid","Tcell","Cd8T"))
FeaturePlot(merged, features=c("Bcell","NK","Stromal","cycling"))
```
Some NK contaminants went through?
```{r fig.height=2.5}
FeaturePlot(merged, features=c("Klrd1","Nkg7","Ncr1","Ncam1","Cd3d","Cd3e"), ncol=3)

FeaturePlot(merged, features=c("Top2a","Mki67","Gzmb","Gzmk","Havcr2","Tcf7"), ncol=3)
```

Expression of typical Tex markers
```{r fig.height=4, fig.width=2}
Idents(merged) <- "functional.cluster"
VlnPlot(merged, features=c("Havcr2","Pdcd1","Lag3","Tox","Gzmb","Gzmk","Tcf7","Ccr7","Mki67","Top2a"), pt.size = 0, stack = T, flip = T)
```


Automatic detection using scGate
```{r}
models <- scGate::get_scGateDB()

cycling.g1 <- scGate::genes.blacklist.default$Mm$cellCycle.G1S
cycling.g2 <- scGate::genes.blacklist.default$Mm$cellCycle.G2M
cycling <- c(cycling.g1, cycling.g2)

gated <- scGate(merged, model = models$mouse$generic$CD8T, additional.signatures = list("cycling"=cycling))
scGate::plot_levels(gated)
```

Potentially ~10% of non-T cells that should be removed (mostly NKs, but also a few myeloid cells).
```{r}
gated <- subset(gated, subset=is.pure=="Pure")
gated
```


```{r}
#Remove sample lung_1 ?
rem_lung1 <- T
if (rem_lung1) {
   gated <- subset(gated, subset = cancer.type != "Lung_1")
}
```

Find DEGs
```{r}
markers <- FindMarkers(gated, group.by = "functional.cluster",  
                       min.diff.pct = 0.2,
                    #   min.pct = 0.5,
                       ident.1 = "CD8_Tex", ident.2 = "CD8_EffectorMemory")

markers <- subset(markers, subset = p_val <= 0.05)
head(markers)
dim(markers)

# Me quedo solo con los markers del volcano (> 0.5 y < -0.5)
markers <- subset(markers, subset = abs(avg_log2FC) >= 0.5)

EnhancedVolcano(markers, x = "avg_log2FC", y = "p_val_adj", lab = rownames(markers))


markers.list <- rownames(markers)
```

Proteasomal genes
```{r}
proteasomal.genes <- c("Psma1","Psma2", "Psma3","Psma4","Psma5","Psma6","Psma7",
                  "Psma8","Psmb1","Psmb2","Psmb3","Psmb4","Psmb5","Psmb6","Psmb7",
                  "Psmb8","Psmb9","Psmb10","Psmc1","Psmc2","Psmc3","Psmc4","Psmc5",
                  "Psmc6","Psmd1","Psmd2","Psmd3","Psmd4","Psmd5","Psmd6","Psmd7",
                  "Psmd8","Psmd9","Psmd10","Psmd11","Psmd12","Psmd13","Psmd14",
                  "Sem1","Adrm1","Psme1","Psme2","Psme3","Psme4","Psmf1")

```

#Generate heatmap

```{r}
DefaultAssay(gated) <- "RNA"
pheatmap <- gated[proteasomal.genes,]

pheatmap_Tex <- subset(pheatmap, subset = functional.cluster == "CD8_Tex")
pheatmap_Tem <- subset(pheatmap, subset = functional.cluster == "CD8_EffectorMemory")

#Subset
set.seed(1234)
pheatmap_Tem <- pheatmap_Tem[, sample(colnames(pheatmap_Tem), size = ncol(pheatmap_Tex), replace=F)]

pheatmap <- merge(pheatmap_Tem, pheatmap_Tex)

DefaultAssay(pheatmap) <- "RNA"

pheatmap <- NormalizeData(pheatmap, normalization.method = "LogNormalize", scale.factor = 10000)
pheatmap <- ScaleData(pheatmap, features = proteasomal.genes)

anno_colors <- list(c("green3", "mediumblue"),
                    c("orange1","snow3", "mediumorchid3","springgreen2"))
             #       c("orange1","snow3", "mediumorchid3","springgreen2","yellow"))

#tiff('Heatmap.tiff', res=700, width = 5000, height = 4000, units='px', compression = 'lzw')
plot_heatmap(dataset = pheatmap,
             row_font_size = 8,
             markers = proteasomal.genes,
             sort_var = c("functional.cluster","cancer.type"),
             anno_var = c("functional.cluster","cancer.type"),
             anno_colors = anno_colors,
             hm_limit = c(-1,0,1),
             hm_colors = c("green3", "white","mediumblue"))
```


######################################################################################
######### GSEA using all genes from the T cells extracted from the 4 studies #########
######################################################################################

# Use all genes from the CD8A expressing T cells from the 4 studies included (withouth lung_1).

```{r}
markers.genes <- wilcoxauc(X = gated, 'functional.cluster')
dplyr::count(markers.genes, group)

markers.genes <- markers.genes %>%
  dplyr::filter(pval <= 0.05) %>%
  dplyr::filter(group == "CD8_Tex") %>%
  arrange(desc(logFC)) %>%
  dplyr::select(feature, logFC)           # select only the features with p< 0.05 and order by logFC.

markers.genes$feature = toupper(markers.genes$feature)

ranks<- deframe(markers.genes)
head(ranks)

barplot(sort(ranks, decreasing = T))
```
# Load the reference

# To do Gene set enrichment analysis, we need to have the annotated gene set first. One popular source is the MsigDB from Broad Institute.
# https://www.gsea-msigdb.org/gsea/msigdb/index.jsp

```{r}
m_df<- msigdbr(species = "Homo sapiens", category = "H")
head(m_df)

fgsea_sets<- m_df %>% split(x = .$gene_symbol, f = .$gs_name)

# Run the analysis

fgseaRes<- fgsea(fgsea_sets, ranks, minSize=15, maxSize = 500, nperm= 1000)

head(fgseaRes[order(padj, -abs(NES)), ], n=10)

#write.xlsx(fgseaRes[,c(1:7)], "GSEA_pathways.xlsx", sheetName = "H_4studies", append = TRUE)

# Plot the results

pathway <- "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY"

#tiff('HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY.tiff', res=700, width = 3000, height = 1300, units='px', compression = 'lzw')
plotEnrichment(fgsea_sets[[pathway]], ranks, ticksSize = 0.5, stats = T) +
  labs(title=pathway) +
  plotGseaTable(fgsea_sets[[pathway]], ranks, fgseaRes, gseaParam = 1) + 
  theme_bw()

#dev.off()


#tiff('Panel GSEA.tiff', res=700, width = 3000, height = 1300, units='px', compression = 'lzw')

dev.off()
```

####################################################
######### Heatmap using reduced gene signature #########
####################################################

```{r}
gene.signature <- c("Cxcr4", "Rbpj", "Rps27a", "Psmb2", "Ldha", "Rbx1", "Psmb8", "Eno1", "Pkm", "Psmc3", "Chchd2", "Pgk1")

pheatmap2 <- gated[gene.signature,]

pheatmap2_Tex <- subset(pheatmap2, subset = functional.cluster == "CD8_Tex")
pheatmap2_Tem <- subset(pheatmap2, subset = functional.cluster == "CD8_EffectorMemory")

pheatmap2_Tem <- pheatmap2_Tem[, sample(colnames(pheatmap2_Tem), size = ncol(pheatmap2_Tex), replace=F)]

pheatmap2 <- merge(pheatmap2_Tem, pheatmap2_Tex)

DefaultAssay(pheatmap2) <- "RNA"

pheatmap2 <- NormalizeData(pheatmap2, normalization.method = "LogNormalize", scale.factor = 10000)
pheatmap2 <- ScaleData(pheatmap2, features = gene.signature)

#tiff('Heatmap_gene signature.tiff', res=700, width = 5000, height = 1500, units='px', compression = 'lzw')

plot_heatmap(dataset = pheatmap2,
             row_font_size = 8,
             markers = gene.signature,
             sort_var = c("functional.cluster","cancer.type"),
             anno_var = c("functional.cluster","cancer.type"),
             anno_colors = list(c("green3", "mediumblue"),
                                c("orange1","snow3", "mediumorchid3","springgreen2")),
             hm_limit = c(-1.5,0,1.5),
             hm_colors = c("green3", "white","mediumblue"))

#dev.off()

violin <- VlnPlot(pheatmap, features = c("Sem1"), group.by = "functional.cluster")
violin$data
#write.xlsx2(violin$data, "human proteasome gene expression-chaperones.xlsx", sheetName = "Psmf1", append= TRUE)

```



